### ISSUE TITLE
Firefox 46 compatibility doesn't work

### ISSUE THREAD

**Issue by gkatsev** [vimperator/vimperator-labs#398]
One of the big compatibility issues that vimperator is going to have is actually with array and generator comprehensions which are being removed in ff46: https://bugzilla.mozilla.org/show_bug.cgi?id=1220564
I've started working at it but it's hard because I don't quite understand how the comprehensions work.

**Comment by reepush** [comment]
> I've started working at it but it's hard because I don't quite understand how the comprehensions work.

I began to experience the same feelings after I saw these lines:
```js
serial: function () [ {
        command: this.name,
        arguments: [abbr.lhs],
        literalArg: abbr.rhs,
        options: abbr.rhs instanceof Function ? {"-javascript": null} : {}
    }
    for ([, abbr] in Iterator(abbreviations.merged))
    if (abbr.modesEqual(modes))
]
```

**Comment by reepush** [comment]
I got the following list of files which use comprehensions:
```
common/content/abbreviations.js
common/content/autocommands.js
common/content/base.js
common/content/buffer.js
common/content/commands.js
common/content/completion.js
common/content/events.js
common/content/hints.js
common/content/io.js
common/content/javascript.js
common/content/liberator.js
common/content/mappings.js
common/content/marks.js
common/content/modes.js
common/content/options.js
common/content/services.js
common/content/statusline.js
common/content/style.js
common/content/util.js
common/modules/template-tag.js
vimperator/content/config.js
```

@gkatsev If you don't mind, I would like to start from the end of the list and help you with it?

**Comment by gkatsev** [comment]
@reepush that would be amazing.

**Comment by SammysHP** [comment]
List comprehensions are a really nice and helpful construction and it's sad that they never were in the standard and that they'll be removed.

> I've started working at it but it's hard because I don't quite understand how the comprehensions work.

The example from above explained:

```JavaScript
serial: function () [ {
        command: this.name,
        arguments: [abbr.lhs],
        literalArg: abbr.rhs,
        options: abbr.rhs instanceof Function ? {"-javascript": null} : {}
    }
    for ([, abbr] in Iterator(abbreviations.merged))
    if (abbr.modesEqual(modes))
]
```

- Create a function `serial` …
- … that returns an array …
- … consisting of objects (`commands`, `arguments`, `literalArg`, `options`).
- An object is created for each element of `abbreviations.merged` …
- … (by creating an Iterator from it, the key->value is decomposed into `nothing` and `abbr`) …
- … **if** `modesEqual()` is true.

Or in other words: An array with an object for every item in `abbreviations.merged` that makes `modesEqual()` true.

**Comment by gkatsev** [comment]
This seems a lot more accessibly and readable to me:
```js
Object.entries(abbreviations.merged)
.filter(function([_, abbr]) { return abbr.modesEqual(modes);})
.map(function([_, abbr]) { return {
  command: this.name,
  arguments: [abbr.lhs],
  literalArg: abbr.rhs,
  options: typeof abbr.rhs === 'function' ? { '-javascript': null} : {}
};
});
```
(Granted Object.entries is es6 but you can do a similar thing with Object.keys and what not).

**Comment by SammysHP** [comment]
I had another look at the deprecation and as far as I understand only the old syntax is removed. See https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Operators/Array_comprehensions

I don't know if the old version is in use somewhere.

**Comment by ecksun** [comment]
We should probably update the HACKING file to not suggest that we use list comprehension:

    Of course, this does not mean, you have to avoid all new JavaScript
    features like list comprehension or generators.

**Comment by reepush** [comment]
@SammysHP @ecksun Thanks, after your comments I finally got that only legacy comprehensions have been removed:
```js
[x for (x in y)]
````

But we can use this one:
```js
[for (x of y) x]
```

This way we can keep our one liners and iterables without moving to arrays.

**Comment by gkatsev** [comment]
I really would prefer to switch away from them altogether. We should not be using non-standard features anymore.

**Comment by SammysHP** [comment]
Possibly it will be in the next version of the standard.

**Comment by gkatsev** [comment]
It was taken *out* of the ES6 standard and it's unlikely to show up any time soon. Even if it does, it doesn't mean it'll be exactly the same syntax.
It's non-standard, that means we shouldn't use it. If we just used standard features we wouldn't be in this mess right now :)

**Comment by SammysHP** [comment]
Hmm… I referred to this comment that says that array comprehensions might come in ES7 or later.
https://bugzilla.mozilla.org/show_bug.cgi?id=1220564#c2

But you're right. Let's stick to standard features.

**Comment by reepush** [comment]
It took quite longer than I expected, but finally it seems to be _almost_ working. Currently there are errors like `ReferenceError: hightlight is not defined` and hints aren't working (likely broken API in 46).

I wrote asserts for testing in 44 to be sure that everything keeps the same, but was unable to trigger some of them. So will look more carefully at these parts, and if someone wants to help with testing – checkout [comprehensions](https://github.com/reepush/vimperator-labs/tree/comprehensions) branch.

**Comment by gkatsev** [comment]
@reepush awesome stuff. Is it ready for a PR and codereview type thing? The `highlight` thing could be related to the `const` and `let` being limited to files nowdays and not being global anymore.

**Comment by reepush** [comment]
@gkatsev Thanks. I want to test and check some assertions again, maybe got feedback from others, and only after that remove commented assertions and file a PR.

**Comment by ecawthon** [comment]
Thanks for working on this, @reepush. I'm testing your [comprehensions](https://github.com/reepush/vimperator-labs/tree/comprehensions) branch, and it looks like mapping doesn't currently work. Steps to reproduce (in both FF 45 stable and 46 beta):
1. I enter`:nnoremap ; :` (or `map ; :` or (`:nnoremap`|`:map`)` j 3j`)
2. At this point, `:messages` is empty
3. I type the mapped key, nothing happens, and the statusline does not update
4. Now `:messages` displays `TypeError: can't convert undefined to object`

**Comment by vaibhavsagar** [comment]
Thank you @reepush! I'm using this on Firefox 46.0b4 and it does almost everything I need except URL autocompletion (search bookmarks work). Is anyone else having the same issue?

**Comment by reepush** [comment]
@ecawthon @vaibhavsagar Thanks for your feedback. Will look at it soon.

**Comment by vaibhavsagar** [comment]
I'm also getting 'TypeError: can't convert undefined to object' when trying to use mappings I've defined.

**Comment by jas01** [comment]
I can't help you guys because I'm not a dev. 
But today I switch (well my OS pkg system) to firefox 46 and I don't getting anything at all. I can event make the ":" show the bottom line.
If anyone want my try something I would glad to help

**Comment by gineadai** [comment]
@jas01  [https://github.com/reepush/vimperator-labs/tree/comprehensions](compehensions) branch work in FF46.

**Comment by jas01** [comment]
@gineadai : Thanks, any chance this "comprehensions" branch would be merge into origin/master soon ? 

Regards

**Comment by gkatsev** [comment]
Hopefully we can get it merged in within the week

**Comment by timss** [comment]
Seems that Github didn't pick up on this, so: #483

**Comment by mpetazzoni** [comment]
Please please get this resolved soon :-) Looks like you're working on it, so that's great. But now that FF46 is being pushed OTA, Vimperator is breaking everywhere! :-(

**Comment by gkatsev** [comment]
Yeah, unfortunately, vimperator can't be a priority right now and we don't have enough people so that it happens anyway. It'll get fixed this week for sure.

**Comment by reepush** [comment]
If someone don't want to build vimperator by himself, [here is a build](http://push.org.ru/grisha/files/vimperator-46.0.xpi.zip) from `comprehensions-review` branch. Don't forget to set `xpinstall.signatures.required` to `false` in _about:config_ and report any bugs you will find here!

And note that it's a _temporary solution_ and official distribution is done through _addons.mozilla.org_ by maintainers 😉

**Comment by gkatsev** [comment]
Finally got a chance to do a quick code review and gave it a whirl. Didn't find anything major and it generally looks good to me. I merged it in and will be making an XPI to upload to mozilla right now.

**Comment by gkatsev** [comment]
Our long national nightmare is over. Version 3.13.0 of vimperator is out and is available on github release: https://github.com/vimperator/vimperator-labs/releases/tag/vimperator-3.13.0. It's been uploaded to Mozilla but it takes them some time to review.

**Comment by gkatsev** [comment]
Also, huge props to @reepush who did the bulk (all?) of the work to get vimperator back up and running with firefox 46!

**Comment by nornagest** [comment]
It works! Thank you so much for your work!

**Comment by rr-** [comment]
Thanks @reepush, I was freaked out when FF pushed 46 upgrade and everything has stopped working, but with your build things are back to normal. Having a binary build is a lifesaver for Windows users.

**Comment by MPW1412** [comment]
Hello,

can you guys give me a little tipp how to force the install of an unsigned addon to firefox 46?

I wanted to try the xpi from the repo here, but firefox doesn't install it.

How did you manage to do so?

**Comment by dkns** [comment]
@MPW1412 Go to about:config and change xpinstall.signatures.required to false

**Comment by MPW1412** [comment]
Thanks, 3.13.0 works like a charm here in Firefox 46.

**Comment by gkatsev** [comment]
It's reviewed and published. Closing as fixed then.
